### PR TITLE
EXTCODEHASH Opcode

### LINKED ISSUE #918: Implement EXTCODEHASH
To use the latest Solidity compiler in default mode (*Constantinople*) EXTCODEHASH should be implemented.

**References**
EIP-1052
https://github.com/ethereum/EIPs/blob/master/EIPS/eip-1052.md

Post-constantinople, we should use extcodehash instead of extcodesize #4834
https://github.com/ethereum/solidity/issues/4834

Solidity Compiler
https://github.com/ethereum/solidity/blob/3d8b323d3b9ebf22f32136ab9f947300fee8c13d/liblangutil/EVMVersion.h#L80-L82

### PR DESCRIPTION
**Pull request by julianlen**
Fixes #918

### CODE REVIEW

**Comment by colltoaction** [comment]
Looking good so far. We'll do a thorough review soon!

**Comment by julianlen** [comment]
Integration test were added [here](https://github.com/rsksmart/rskj-truffle-tests/pull/1)

File: rskj-core/src/test/java/co/rsk/db/MutableTrieCacheTest.java [hunk 1]
```diff
@@ -323,4 +324,79 @@ public void testStorageKeysDeletedAccount() {
...
+        baseMutableTrie.put(key, value);
+        byte[] result = baseMutableTrie.get(key);
+        Assert.assertArrayEquals(value, result);
+
+        Uint24 valueLength = baseMutableTrie.getValueLength(key);
+        Assert.assertEquals(valueLength, new Uint24(value.length));
+
+        // Test the same in cache
+
+        MutableTrieCache mtCache = new MutableTrieCache(baseMutableTrie);
```
Comment (diega) [inline rskj-core/src/test/java/co/rsk/db/MutableTrieCacheTest.java]: I would move the `co.rsk.db.MutableTrieCache` tests to its own class. Or maybe we can parametrize the test receiving two `co.rsk.trie.MutableTrie` implementations, one `co.rsk.db.MutableTrieCache` and one `co.rsk.db.MutableTrieImpl`.

File: rskj-core/src/test/java/org/ethereum/config/blockchain/upgrades/ActivationConfigTest.java [hunk 2]
```diff
@@ -60,6 +61,7 @@
             "    rskip126: wasabi100",
             "    rskip132: wasabi100",
             "    rskip136: bahamas",
+            "    rskip140: wasabi120",
```
Comment (diega) [inline rskj-core/src/test/java/org/ethereum/config/blockchain/upgrades/ActivationConfigTest.java]: Besides the _wasabi120_ reference, I don't think this test needs to be modified on every new _network upgrade_, I think it's testing the features avaiable through the configuration mechanism, not and exhaustive existence of every _network upgrade_ definition.

File: rskj-core/src/main/java/co/rsk/db/MutableTrieImpl.java [hunk 3]
```diff
@@ -86,6 +86,15 @@ public Uint24 getValueLength(byte[] key) {
         return atrie.getValueLength();
     }
 
+    @Override
+    public Keccak256 getValueHash(byte[] key) {
+        Trie atrie = trie.find(key);
+        if (atrie == null) {
+            return Keccak256.ZERO_HASH;
```
Comment (diega) [inline rskj-core/src/main/java/co/rsk/db/MutableTrieImpl.java]: I think a reference here to the EIP stating that missing keys should hash to _zero_ will be greatly appreciated

File: rskj-core/src/main/java/org/ethereum/vm/VM.java [hunk 4]
```diff
@@ -788,6 +790,42 @@ protected void doCODESIZE() {
...
+            if (codeHash.equals(Keccak256.ZERO_HASH)) {
+                program.stackPush(DataWord.ZERO);
+            } else {
+                DataWord word = DataWord.valueOf(codeHash.getBytes());
+                program.stackPush(word);
+            }
+        }
+
+        if (isLogEnabled) {
+            hint = "hash: " + codeHash;
```
Comment (diega) [inline rskj-core/src/main/java/org/ethereum/vm/VM.java]: Is it possible that, in the case of a precompiled contract, the `codeHash` value stored in the _hint_ is wrong?

File: rskj-core/src/test/java/org/ethereum/vm/program/invoke/ProgramInvokeMockImpl.java [hunk 5]
```diff
@@ -68,6 +68,7 @@ public ProgramInvokeMockImpl(byte[] contractCode, RskAddress contractAddress) {
         //Defaults to defaultContractAddress constant defined in this mock
         this.contractAddress = contractAddress!=null?contractAddress:this.defaultContractAddress;
         this.repository.createAccount(this.contractAddress);
+        this.repository.setupContract(this.contractAddress);
```
Comment (diega) [inline rskj-core/src/test/java/org/ethereum/vm/program/invoke/ProgramInvokeMockImpl.java]: good catch!

File: rskj-core/src/test/java/org/ethereum/config/blockchain/upgrades/ActivationConfigTest.java (hunk 2 above)
Comment (julianlen) [reply rskj-core/src/test/java/org/ethereum/config/blockchain/upgrades/ActivationConfigTest.java]: The test should be modified to make it pass

File: rskj-core/src/main/java/co/rsk/db/MutableTrieImpl.java (hunk 3 above)
Comment (julianlen) [reply rskj-core/src/main/java/co/rsk/db/MutableTrieImpl.java]: Actually RSKIP140 or [EIP](https://eips.ethereum.org/EIPS/eip-1052) specifies that if an account does not exist, zero is returned. At MutableTrieImpl level ZERO_HASH is a way to difference between Account/Contract and Non-existing account.

Comment (julianlen) [reply rskj-core/src/main/java/co/rsk/db/MutableTrieImpl.java]: BTW, i wrote this comment: _From RSKIP140 empty-account specification._

File: rskj-core/src/main/java/org/ethereum/vm/VM.java (hunk 4 above)
Comment (julianlen) [reply rskj-core/src/main/java/org/ethereum/vm/VM.java]: Yes, that was a problem, it was solved [here](https://github.com/rsksmart/rskj/blob/6833151421e39bcc84b4a37ef594e32a75b02699/rskj-core/src/main/java/org/ethereum/vm/VM.java#L810)

File: rskj-core/src/test/java/co/rsk/db/MutableTrieCacheTest.java (hunk 1 above)
Comment (julianlen) [reply rskj-core/src/test/java/co/rsk/db/MutableTrieCacheTest.java]: > This looks great, I just made minor comments.
> I think `9373882647277a3273aae53ea2597f5281e60d11` can be fixed up into `6c1fa26dae6086916a02e92b8ee5bcfdda9cfca6`.
> Also, are the [test cases described in EIP-1052](https://github.com/ethereum/EIPs/blob/master/EIPS/eip-1052.md#test-cases) covered by the _integration tests_ suite?

Yes, you can check it [here](https://github.com/rsksmart/rskj-truffle-tests/pull/1/commits/bdfc52b6305ccd8c40f6f94d23cacee833158ba5), each test has which case is testing.

Comment (julianlen) [reply rskj-core/src/test/java/co/rsk/db/MutableTrieCacheTest.java]: About co.rsk.db.MutableTrieCache tests, i made some [changes](https://github.com/rsksmart/rskj/blob/2f8340b236b5508ee07875bf8cb8ba7488c560be/rskj-core/src/test/java/co/rsk/db/MutableTrieCacheTest.java#L406)

File: rskj-core/src/main/java/org/ethereum/vm/VM.java [hunk 6]
```diff
@@ -788,6 +790,48 @@ protected void doCODESIZE() {
...
+        }
+
+        //EXECUTION PHASE
+        DataWord address = program.stackPop();
+
+        ActivationConfig.ForBlock activations = program.getActivations();
+        PrecompiledContracts.PrecompiledContract precompiledContract = precompiledContracts.getContractForAddress(activations, address);
+        boolean isPrecompiledContract = precompiledContract != null;
+
+        if (isPrecompiledContract) {
```
Comment (colltoaction) [inline rskj-core/src/main/java/org/ethereum/vm/VM.java]: How about

```java
if (isPrecompiledContract) {
    ...
} else {
    Keccak256 codeHash = program.getCodeHashAt(address);
    ...
}

program.step();
```

File: rskj-core/src/main/resources/config/regtest.conf [hunk 7]
```diff
@@ -9,6 +9,7 @@ blockchain.config {
     },
     consensusRules = {
         rskip97 = -1 # disable orchid difficulty drop
+
```
Comment (colltoaction) [inline rskj-core/src/main/resources/config/regtest.conf]: Please undo this change

File: rskj-core/src/main/java/org/ethereum/db/MutableRepository.java [hunk 8]
```diff
@@ -156,6 +157,21 @@ public synchronized int getCodeLength(RskAddress addr) {
         return mutableTrie.getValueLength(key).intValue();
     }
 
+    public synchronized Keccak256 getCodeHash(RskAddress addr) {
```
Comment (colltoaction) [inline rskj-core/src/main/java/org/ethereum/db/MutableRepository.java]: Missing `@Override`

File: rskj-core/src/main/resources/config/regtest.conf (hunk 7 above)
Comment (julianlen) [reply rskj-core/src/main/resources/config/regtest.conf]: Solved!

[the remark this answers is not in the document]
Comment (julianlen) [reply rskj-core/src/main/resources/config/regtest.conf]: Solved!

File: rskj-core/src/test/java/co/rsk/db/MutableTrieCacheTest.java [hunk 9]
```diff
@@ -323,4 +326,131 @@ public void testStorageKeysDeletedAccount() {
...
+
+        // First put some strings in the base
+
+        byte[] value = toBytes("alice");
+        byte[] key = toBytes("ALICE");
+        byte[] keyForCache = toBytes("ALICE2");
+
+        baseMutableTrie.put(key, value);
+        byte[] result = baseMutableTrie.get(key);
+        Assert.assertArrayEquals(value, result);
```
[the remark this answers is not in the document]
Comment (julianlen) [reply rskj-core/src/test/java/co/rsk/db/MutableTrieCacheTest.java]: I am making the changes in the commit "getValueLength tests added".

[the remark this answers is not in the document]
Comment (julianlen) [reply rskj-core/src/test/java/co/rsk/db/MutableTrieCacheTest.java]: solved, the changes were mainly in "getValueLength test"

File: rskj-core/src/test/java/co/rsk/db/MutableTrieCacheTest.java [hunk 10]
```diff
@@ -323,4 +326,131 @@ public void testStorageKeysDeletedAccount() {
...
+        MutableTrieImpl baseMutableTrie = new MutableTrieImpl(null, new Trie());
+        MutableTrieCache mtCache = new MutableTrieCache(baseMutableTrie);
+
+        byte[] wrongKey = toBytes("BOB");
+
+        executeHashTest(baseMutableTrie, wrongKey, null, Keccak256.ZERO_HASH);
+        executeHashTest(mtCache, wrongKey, null, Keccak256.ZERO_HASH);
+    }
+
+    private void executeHashTest(MutableTrie trie, byte[] key, byte[] value, Keccak256 expectedHash) {
```
[the remark this answers is not in the document]
Comment (julianlen) [reply rskj-core/src/test/java/co/rsk/db/MutableTrieCacheTest.java]: done!

File: rskj-core/src/main/java/org/ethereum/db/MutableRepository.java (hunk 8 above)
Comment (julianlen) [reply rskj-core/src/main/java/org/ethereum/db/MutableRepository.java]: fixed!

File: rskj-core/src/main/java/org/ethereum/vm/VM.java (hunk 6 above)
Comment (julianlen) [reply rskj-core/src/main/java/org/ethereum/vm/VM.java]: fixed!

File: rskj-core/src/main/java/org/ethereum/vm/VM.java [hunk 11]
```diff
@@ -788,6 +790,42 @@ protected void doCODESIZE() {
...
+            program.stackPush(DataWord.valueOf(Keccak256Helper.keccak256(EMPTY_BYTE_ARRAY)));
+        }
+
+        Keccak256 codeHash = program.getCodeHashAt(address);
```
[the remark this answers is not in the document]
Comment (julianlen) [reply rskj-core/src/main/java/org/ethereum/vm/VM.java]: fixed!

File: rskj-core/src/test/java/co/rsk/vm/ExtCodeHashTest.java [hunk 12]
```diff
@@ -0,0 +1,140 @@
...
+
+    private static Transaction createTransaction() {
+        int number = 0;
+        AccountBuilder acbuilder = new AccountBuilder();
+        acbuilder.name("sender" + number);
+        Account sender = acbuilder.build();
+        acbuilder.name("receiver" + number);
+        Account receiver = acbuilder.build();
+        TransactionBuilder txbuilder = new TransactionBuilder();
+        return txbuilder.sender(sender).receiver(receiver).value(BigInteger.valueOf(number * 1000 + 1000)).build();
```
[the remark this answers is not in the document]
Comment (julianlen) [reply rskj-core/src/test/java/co/rsk/vm/ExtCodeHashTest.java]: fixed!
